codex/utils/timer.nim

```diff
@@ -0,0 +1,54 @@
+## Nim-Codex
+## Copyright (c) 2023 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+## Timer
+## Used to execute a callback in a loop
+
+import pkg/chronos
+import pkg/chronicles
+import pkg/questionable
+import pkg/upraises
+
+type
+  TimerCallback* = proc(): Future[void] {.gcsafe, upraises:[].}
+  Timer* = ref object of RootObj
+    callback: TimerCallback
+    interval: Duration
+    name: string
+    loopFuture: ?Future[void]
+
+proc new*(T: type Timer, timerName = "Unnamed Timer"): T =
+  T(
+    name: timerName,
+    loopFuture: Future[void].none
+  )
+
+proc timerLoop(timer: Timer) {.async.} =
+  try:
+    while true:
+      await sleepAsync(timer.interval)
+      await timer.callback()
+  except Exception as exc:
```

Comment:
`Exception` is the root of the exception hierarchy in Nim, including fatal errors (ouch!), usually anything except `CatchableError` should be left alone.

Reply:
Changed it to CatchableError.
Is it true that because this is running in a Future, anything not caught here will cause the application to crash?

Reply:
> Is it true that because this is running in a Future, anything not caught here will cause the application to crash?

It depends. 

If you `asyncSpawn` the future any exception is propagated freely without any chance of being caught, so any spawned proc should guard itself with a `try/except CatchableError ...`. 

Futures that are awaited or stored in a variable, can be handled normally with a `try/except/finally` as you would expect.

so:

```nim
proc throws(): Future[int] {.async.} = 
  raise newException(CatchableError, "Error!")

# 1) handling exceptions and propagating cancellations
try:
  discard await throws()
except CatchableError as exc:
  # do something
except CancelledError as exc:
   raise  exc

# 2) store future in variable
let fut = throws() 

# same as `1`
try:
  discard await fut
except CatchableError as exc:
  # do something
except CancelledError as exc:
   raise  exc

# same as `1` but reading the value from the future, will reaise the stored exception if it failed (same as `await`)
try:
  discard fut.read()
except CatchableError as exc:
  # do something
except CancelledError as exc:
   raise  exc

# 3) `asyncSpawn` that crashes the app, nothing the catch the exception
asynSpawn throws()

# 4) `asyncSpawn` that doesn't crash (should not propagate cancellations!)
proc safeSpawn() {.async.} =
  try:
    discard await throws()
  except CatchableError as exc:
    # do something without crashing

asyncSpawn safeSpawn()

```

Cancellation should almost always be propagated, unless they are in an async spawn (or similar code).

Reply:
This makes sense. Thanks. It's a good reason to prefer storing futures in variables.